libyul/optimiser/MemoryEscalator.cpp

```diff
@@ -0,0 +1,137 @@
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/StackToMemoryMover.h>
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/AsmData.h>
+#include <libyul/CompilabilityChecker.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+#include <libyul/Object.h>
+#include <libyul/Utilities.h>
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/CommonData.h>
+#include <list>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+
+void MemoryEscalator::run(OptimiserStepContext& _context, Object& _object, bool _optimizeStackAllocation)
+{
+	if (!_object.code)
+		return;
+
+	// Determine which variables need to be moved.
+	auto functionStackErrorInfo = CompilabilityChecker::run(_context.dialect, _object, _optimizeStackAllocation);
+	if (functionStackErrorInfo.empty())
+		return;
+
+	// Find the literal argument of the ``memoryinit`` call, if there is a unique such call, otherwise abort.
+	yulAssert(dynamic_cast<EVMDialect const*>(&_context.dialect), "MemoryEscalator can only be run on EVMDialect objects.");
+	Literal* memoryInitLiteral = nullptr;
+	if (
+		auto memoryInits = FunctionCallFinder::run(*_object.code, "memoryinit"_yulstring);
+		memoryInits.size() == 1
+	)
+		memoryInitLiteral = std::get_if<Literal>(&memoryInits.front()->arguments.back());
+	if (!memoryInitLiteral)
+		return;
+	u256 reservedMemory = valueOfLiteral(*memoryInitLiteral);
+
+	auto callGraph = CallGraphGenerator::callGraph(*_object.code).functionCalls;
+
+	// Collect all names of functions contained in cycles in the callgraph.
+	// TODO: this algorithm is suboptimal and can be improved.
+	std::set<YulString> containedInCycle;
+	auto findCycles = [
+		&,
+		visited = std::map<YulString, uint64_t>{},
+		currentPath = std::vector<YulString>{}
+	](YulString _node, auto& _recurse) mutable -> void
+	{
+		if (auto it = std::find(currentPath.begin(), currentPath.end(), _node); it != currentPath.end())
+			containedInCycle.insert(it, currentPath.end());
+		else
+		{
+			visited[_node] = currentPath.size();
+			currentPath.emplace_back(_node);
+			for (auto const& child: callGraph[_node])
+				_recurse(child, _recurse);
+			currentPath.pop_back();
+		}
+	};
+	findCycles(YulString{}, findCycles);
+
+	using SlotAllocations = map<YulString, map<YulString, u256>>;
+	SlotAllocations slotAllocations;
+	uint64_t requiredSlots = 0;
+
+	// Walks the call graph using a Depth-First-Search.
+	// The visit assigns slots for the variables that need to be moved in the current function.
+	// Each visit resets the number of used slot at the end, resulting in all child visits starting
+	// at the same number of reserved slots.
+	// Since within cycles there are no variables to be moved, this will result in each variable having a unique
+	// offset on each path through the call graph. However, distinct paths may share the same offsets for different
+	// variables.
+	auto allocateMemorySlots = [
```

Comment:
This is a rather critical produce. Is there a way to have it better isolated so that we can write unit-tests for it? I'm thinking about e.g. call graphs of the form
```
{
  f()
  g()
  function f() { h() }
  function g() { h() }
  function h() {}
}
```

Reply:
Yeah... maybe unit tests in which simply *every* variable is marked for escalating would be way better to verify this than to construct actual test cases - I'll see if I can extract this thing into yet another standalone step and unit test it.

Reply:
Actually we might also add a routine that just verifies that the allocations do not conflict by forwarding all allocations to the callees - but that might turn out to be the same algorithm after all...